### PR TITLE
VH(bb) analyzer development: JER + btag weights

### PR DESCRIPTION
**Pull request by scooperstein**
Hi @arizzi ,

Please don't merge. I have included the btag weights module. Disregard what I said on Friday, looks like it is working just fine. JEC also fine. I have tried to include also the JER smearing but am having problems. When I run the VH(bb) post-processor with this implementation I get the output below. Maybe I just don't have the correct syntax to include the JER? Is there somewhere I could reference for the syntax?

Thanks,
Stephane

Loading btagSF weights for cMVA algorithm from file '/afs/cern.ch/work/s/scoopers/private/nanoAOD_dev/CMSSW_9_4_0_pre3/src/PhysicsTools/NanoAODTools/data/btagSF/btagSF_cMVAv2_ichep2016.csv'
Load Library 'CondFormatsBTauObjects'
Load Library 'CondToolsBTau'
Will write selected trees to .
Loading jet energy scale (JES) uncertainties from file '/afs/cern.ch/work/s/scoopers/private/nanoAOD_dev/CMSSW_9_4_0_pre3/src/PhysicsTools/NanoAODTools/data/jme/Summer16_23Sep2016V4_MC_UncertaintySources_AK4PFchs.txt'
%MSG-e JetCorrectorParameters: 
No definitions found!!!
%MSG
%MSG-e JetCorrectorParameters: 
the requested section  doesn't exist!
%MSG

 *** Break *** segmentation violation

....
...
...

Traceback (most recent call last):
  File "postproc.py", line 40, in <module>
    p.run()
  File "/afs/cern.ch/work/s/scoopers/private/nanoAOD_dev/CMSSW_9_4_0_pre3/python/PhysicsTools/NanoAODTools/postprocessing/framework/postprocessor.py", line 54, in run
    for m in self.modules: m.beginJob()
  File "/afs/cern.ch/work/s/scoopers/private/nanoAOD_dev/CMSSW_9_4_0_pre3/python/PhysicsTools/NanoAODTools/postprocessing/modules/jme/jetmetUncertainties.py", line 52, in beginJob
    self.jesUncertainty = ROOT.JetCorrectionUncertainty(os.path.join(self.jesInputFilePath, self.jesUncertaintyInputFileName))
TypeError: none of the 3 overloaded methods succeeded. Full details:
  JetCorrectionUncertainty::JetCorrectionUncertainty() =>
    takes at most 0 arguments (1 given)
  JetCorrectionUncertainty::JetCorrectionUncertainty(const string& fDataFile) =>
    problem in C++; program state has been reset
  JetCorrectionUncertainty::JetCorrectionUncertainty(const JetCorrectorParameters& fParameters) =>
    could not convert argument 1

### CODE REVIEW

File: python/postprocessing/analysis/higgs/vhbb/postproc.py [hunk 1]
```diff
@@ -34,5 +36,5 @@
 Sum$(Jet_pt *(abs(Jet_eta)<2.5 && Jet_pt > 20 && Jet_jetId)) > 160  || 
 MET_pt > 100  || Sum$(Muon_pt > 20 && Muon_tightId) >= 1
 '''
-p=PostProcessor(".",files,selection.replace('\n',' '),"keep_and_drop.txt",[btagSFProducer("cmva"),jecUncertAll_cppOut(),vhbb()],provenance=True)
+p=PostProcessor(".",files,selection.replace('\n',' '),"keep_and_drop.txt",[jecUncertAll_cppOut(),jetmetUncertaintiesAll(),btagSFProducer("cmva"),vhbb()],provenance=True)
```
Comment (scooperstein) [inline python/postprocessing/analysis/higgs/vhbb/postproc.py]: @arizzi specifically I imagine it is this line that perhaps I am not doing correctly to include the JER

Comment (arizzi) [reply python/postprocessing/analysis/higgs/vhbb/postproc.py]: it seems there is a jetSmearer module from @veelken can you try that?

Comment (scooperstein) [reply python/postprocessing/analysis/higgs/vhbb/postproc.py]: right the jetSmearer module is an integral part of this jetmetUncertainties module. My understanding from looking at the code for these modules is that one should call the jetmetUncertainties module which wraps around jetSmearer but maybe I have that wrong...

**Comment by arizzi** [comment]
@veelken do you have a working example for stephane to test?

**Comment by veelken** [comment]
Hi,

a fully working example for smearing jets and MET on PAT-tuple/miniAOD level is this tool, maintained by the JetMET POG: https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuidePATTools#MET_Systematics_Tools . 
I am using the nanoAOD version of this tool tool: https://github.com/cms-nanoAOD/nanoAOD-tools/blob/master/python/postprocessing/modules/jme/jetmetUncertainties.py .

I hope this helps,


Christian

**Comment by arizzi** [comment]
I meant a working example in nanoAOD-tools

On Tue, Jan 16, 2018 at 4:02 PM, Christian Veelken <notifications@github.com

**Comment by veelken** [comment]
Hmh, just run the jetmetUncertaintiesProducer module ?

**Comment by scooperstein** [comment]
Thanks Andrea and Christian for your help. Christian I tried to just include the jetmetUncertaintiesProducer module as follows in the post-processor

+p=PostProcessor(".",files,selection.replace('\n',' '),"keep_and_drop.txt",[jecUncertAll_cppOut(),jetmetUncertaintiesAll(),btagSFProducer("cmva"),vhbb()],provenance=True)

but I get the errors listed in the description for this pull request. Sorry it is probably something very simple to fix, but have you done something similar and would you have some working example I could reference? Thanks.

**Comment by scooperstein** [comment]
I notice now that jetmetUncertaintiesAll() fails as I have been saying but jetmetUncertainties() module appears to work fine.

**Comment by scooperstein** [comment]
Upon further investigation it seems to me like the jetmetUncertainties module is meant to calculate the JEC and JER, whereas the jetUncertAll_cppOut module is now deprecated and was only meant for JEC. Additionally it seems that the implementation of loading the JEC library in jetmetUncertainties was not working for the factorized JEC. I have rewritten that to mimic what was in Heppy and what is in jetUncertainties. It looks that everything is now working as expected but it would be nice to confirm with @veelken and @arizzi . Thanks

**Comment by arizzi** [comment]
jetUncertAll_cppOut is definitelly not deprecated, we actually need to compute also JER in C++
I've no time this week to look into it but I will soon

**Comment by scooperstein** [comment]
@arizzi okay then I am missing something, it looks to me like the JEC evaluation is implemented in both jetmetUncertainties.py and jecUncertainties.py so it is not clear to me why one would run both. Thanks for your help, please let me know of any developments. It looks to me like the code as in this PR is including both the JEC and JER properly now but perhaps it is inefficient since the evaluation is all in python now (I think).

**Comment by arizzi** [comment]
@veelken why is the metUncertainty producer also recomputing jet uncertainty? Wouldn't it be better to just apply there the uncertainty computed from the previous module?

**Comment by arizzi** [comment]
@scooperstein exactly, you can go ahead with it for VHbb for now but we have to fix the implementations

**Comment by scooperstein** [comment]
@arizzi Okay thanks then I will go ahead with the implementation as in this PR for VH(bb) for the  time being. I would like to run this on a comprehensive set of nanoAOD 2016 samples so that VH analyzers can try an A->Z 2016 analysis starting from nanoAOD. Do you expect this implementation to be sufficient for that?

**Comment by veelken** [comment]
Hi Andrea,

the jetmetUncertainties.py tool needs to compute the jet energy scale and resolution uncertainties centered on the smeared jet energy and then propagate these differences to the MET. The corresponding PAT-tuple/miniAOD tool https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuidePATTools#MET_Systematics_Tools does it this way, and I have converted this tool to the nanoAOD module framework. If you can think of a better (faster) way of doing this computation in C++, I am all for it, but we would need to make sure that everything is always consistent if we compute the values across different python/C++ codes.

**Comment by arizzi** [comment]
so perhaps we should first smear, then compute (in c++) the uncertainties?
If that is the right procedure for MET than it must be the right procedure also for downstream analysis..

**Comment by arizzi** [comment]
@rappoccio is also looking in the jme nanoaodtools

(sal, this is the discussion I was mentioning in skype)

**Comment by rappoccio** [comment]
Thanks. I actually asked Mikko about this, indeed we should first smear and then compute the uncertainties.

I have a newer implementation of this here:

https://github.com/rappoccio/nanoAOD-tools/tree/TTbarResHad/python/postprocessing/modules/jme

This also adds the JMS/JMR smearing as well as the data/MC scale factor for JMS (still have to add the JMS MC corrections from Thea). The current corrections are trivial, thus so is the implementation, but I think we (JME) needs to put this into the official database framework like the JECs and JERs. 

A full use case for an example can be found here:

https://github.com/rappoccio/nanoAOD-tools/blob/TTbarResHad/python/postprocessing/analysis/b2g/ttbarres/TTbarResAnaHadronic.py

Unfortunately, this is all very, very slow. I can only get <100-200Hz processing speeds. This is about as optimized as I can get it within python, so we definitely need a C++ implementation. I'm looking into that now. It may also be necessary to optimize the JEC uncertainty code itself, which has not been substantively optimized since ~2009.  

If anyone is also willing to help we can divide the tasks.

**Comment by scooperstein** [comment]
hi everyone, just to add one point to this: it would be very nice if in the JES/JER implementation we also store in the post-processed nanoAOD ntuple the applied JES and JER correction separately, as was previously done in Heppy. This would be very helpful for some studies/validation checks. Thanks.

**Pull request merged**